Improve keybinding JSON schema

davidwin opened the pull request:
## Summary of the Pull Request
The pattern regex in [profiles.schema.json](https://raw.githubusercontent.com/davidwin/terminal/keychords_schema/doc/cascadia/profiles.schema.json) now correctly disallows keybindings consisting of only modifiers, modifiers not separated by `+`, and unknown key names. Certain `shift+numpad` combinations are also not allowed.

The description lists allowed key names in tabular format (assuming the client renders `\t` correctly):
![image](https://user-images.githubusercontent.com/9844625/70371759-8fee0f00-18d7-11ea-85c2-b9f739252129.png)

## PR Checklist
* [x] Closes #3856 and #2967
* [x] CLA signed.
* [ ] Tests added/passed _(only manual validation in VSCode)_
* [ ] Requires documentation to be updated _(this is documentation)_
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #3856

## Detailed Description of the Pull Request / Additional comments
The problem with the current keybinding schema is primarily that it accepts invalid key combinations.
Also, there is no accessible documentation of valid keys and modifiers. Now the regular expression 
should be much closer to what the parser in [KeyChordSerialization.cpp](https://github.com/microsoft/terminal/blob/master/src/cascadia/TerminalApp/KeyChordSerialization.cpp) actually accepts, and the associated description explains the allowed string format.

One possible difference between the regexp and the parser is that the former accepts any single character except `+` and whitespace (e.g. `}`) for the key part while the latter relies on the `VkKeyScanW` Windows API for validation of single-character keys.

The regular expression might be a bit difficult to digest, so here it is with some added whitespace and no JSON string escapes:
```regexp
^
 (?<modifier> (ctrl|alt|shift) 
              (?:\+(ctrl|alt|shift)(?<!\2))? 
              (?:\+(ctrl|alt|shift)(?<!\2|\3))?
              \+
 )?
 (?<key> [^\s+]
         |backspace|tab|enter|esc|escape|space|pgup|pageup|pgdn|pagedown|end|home|left|up|right|down|insert|delete
         |(?<!shift.+)(?:numpad_?[0-9]|numpad_(?:period|decimal))
         |numpad_(?:multiply|plus|minus|divide)
         |f[1-9]|f1[0-9]|f2[0-4]
         |plus
 )
$
```
**Q**: Is named captures a good idea? It was used by the previous regular expression, but an alternative approach would be to use numbered captures and only capture the individual modifiers (1 to 3) and the key (4).

**Q**: Would it be better to not list key names _and_ descriptions in an attempted table, which currently isn't rendered very nicely in VS Code? The names are, after all, rather self explanatory.

## Validation Steps Performed
I have performed validation of the new schema in VS Code v1.43 by opening the `profiles.json` settings file and changing the `$schema` property to https://raw.githubusercontent.com/davidwin/terminal/keychords_schema/doc/cascadia/profiles.schema.json. I'm then able to see that I get a description of how to format the `globals.keybingings[].keys` strings if I hover over a `keys` item, and a yellow squiggly if I enter an invalid key combination. All the default key combinations pass without error, and examples of strings that don't validate are: `alt`, `shift+`, `ctrl++`, `+`, `shiftalt+up`.

You can try the regular expression and get a deeper understanding of it at [Regex101](https://regex101.com/r/mqlYhr/4).

DHowett-MSFT commented:
(I'm excited about this; sorry I haven't yet reviewed it!)

davidwin commented:
No problem! I put the PR in draft anyway because I had the two open questions about the style of the regex and the (long) descriptive text.

miniksa commented:
> **Q**: Is named captures a good idea? It was used by the previous regular expression, but an alternative approach would be to use numbered captures and only capture the individual modifiers (1 to 3) and the key (4).

I personally like named captures, but if we're not actually leveraging the names anywhere to consume the groups... then it's probably unnecessary.

> **Q**: Would it be better to not list key names _and_ descriptions in an attempted table, which currently isn't rendered very nicely in VS Code? The names are, after all, rather self explanatory.

I don't think we need to have a description that is pretty much the same thing as a name. If it's cleaner with just the names, I think we just use the names./

oising commented:
When I wrote this, it was always intended for a collaborative improvement session and we had a few nice tweaks on Twitter. I had thought about using explicit names for all of the keys, but figured it might have been a bit futile when you take multilingual/international keyboards into account, so I didn't bother. So be warned - this may block certain combinations that are quite legal to assign outside of the USA. 

As for the named capture groups, I put them in more for pseudo-documentation purposes.

DHowett-MSFT commented:
I'll wait one day before marking this ready for review (so that it can be merged)

davidwin commented:
Great! I noticed that some strings had been added to KeyChordSerialization.cpp since I opened this PR. Just let me rebase the file and add those tweaks, and I'll mark it ready for review tomorrow.

DHowett-MSFT commented:
Also, really sorry for the delay :(

davidwin commented:
I rebased the PR and updated the pattern to match the current `KeyChordSerialization.cpp` (some synonyms had been added). I also made the descriptive text somewhat more consistent.

And don't worry about the delay!

cinnamon-msft commented:
I'm going to leave this one up to @carlos-zamora. He's the resident regex wizard :)
